app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -0,0 +1,251 @@
+@SuppressWarnings("checkstyle:ClassFanOutComplexity")
+//  1)  This warning is because the class relies on too many external classes, which can't really be avoided
+public class QRExchangeKeyActivity extends AppCompatActivity {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("QRExchangeKeyActivity");
+
+    private IServiceLocator serviceLocator;
+
+    private AlertDialog lastDialog;
+
+
+    private IKey publicKey;
+
+    /**
+     * Creates the new activity, should only be called by Android
+     *
+     * @param savedInstanceState Don't touch this
+     */
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        this.setContentView(R.layout.activity_qrexchange);
+
+        final Intent intent = getIntent();
+        this.serviceLocator = (IServiceLocator) intent.getSerializableExtra(ConstantKeywords.SERVICE_LOCATOR);
+
+
+        //TODO: Get the user's generated public key from the database
+        final IKeyGenerator keyGenerator = serviceLocator.getKeyGenerator();
+        final KeyPair pair = keyGenerator.generateRSAKeyPair("test");
+        publicKey = pair.getPublicKey();
+
+
+        final Button scanButton = (Button) findViewById(R.id.scanbutton);
+        scanButton.setOnClickListener(new View.OnClickListener() {
+
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started scanning QR code");
+                final IntentIntegrator integrator = new IntentIntegrator(QRExchangeKeyActivity.this);
+                integrator.initiateScan();
+
+            }
+        });
+
+        final Button generateButton = (Button) findViewById(R.id.generateQRbutton);
+        generateButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started generating QR code");
+                final String space = " ";
+                final Bitmap qrCode = QRGenerator.encode(publicKey.getType() + space + publicKey.getName()
+                        + space + publicKey.getKey());
+                showQRCode(qrCode);
+            }
+        });
+
+        final ImageButton backButton = (ImageButton) findViewById(R.id.backButtonQRExchange);
+        backButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                finish();
+            }
+        });
+
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onActivityResult(final int requestCode, final int resultCode, final Intent intent) {
+        final IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
+        if (scanResult != null) {
+            final String result = scanResult.getContents();
+            addNewContact(result);
+
+        } else {
+            LOGGER.error("No scan result in QR Scanner");
```

Comment:
You can better use LOGGER.error("oops, asdfasdfasdf went wrong", e), because then the log will contain both the stacktrace And the message automatically (and also a beautiful message that asdfasdfasdf went wrong)